Show Last Known Location in LocationPickerActivity

devarsh-mavani-19 opened the pull request:
**Description (required)**
In the LocationPickerActivity The location should be the last used location. It was showing a location near san fransico. So changed it to use Last Picked Location

Fixes #4777 

What changes did you make and why?
In the MediaDetailFragment, I have added an else if branch that gets the last location from ApplicationKvStore (Only if location was not already set by image uploader)

**Tests performed (required)**
Tested on Redmi 6 PRO
Build Variant BETA
API Level 28

**Screenshots (for UI changes only)**
![WhatsApp Image 2022-02-14 at 12 08 26 (1)](https://user-images.githubusercontent.com/55136047/153812984-95bb81cc-f4a6-437b-ba32-f91c6c659d94.jpeg)
![WhatsApp Image 2022-02-14 at 12 08 26](https://user-images.githubusercontent.com/55136047/153812998-05f24333-44e0-4df5-94d0-3d57d168e844.jpeg)

devarsh-mavani-19 commented:
@nicolas-raoul I have created PR. It is showing a failed test and 2 passed tests. Is this normal for small changes?

nicolas-raoul reviewed (changes requested):
Would it be possible to add a unit test that goes inside this `else if` condition?
Thanks a lot!

devarsh-mavani-19 commented:
> Would it be possible to add a unit test that goes inside this `else if` condition? Thanks a lot!

@nicolas-raoul I am not sure how to cover else if condition using testing I explored the MediaDetailFragmentUnitTests class to look for similar tests to mine. I think in the test method I have to set the coordinates using mockito and then call fragment.onActivityResult and pass intent containing values. is this the correct way of doing it? I am new to mockito so I am finding it difficult. Can you help me or give me an idea?
 Thanks

nicolas-raoul commented:
Everyone: Can someone please help? I am also unfamiliar with this. 😅

devarsh-mavani-19 commented:
@nicolas-raoul Finally I have written test cases and passed the code coverage tests as well. Please can you review it.

madhurgupta10 commented on app/src/test/kotlin/fr/free/nrw/commons/media/MediaDetailFragmentUnitTests.kt:
```diff
@@ -229,6 +231,22 @@ class MediaDetailFragmentUnitTests {
+    fun testOnUpdateCoordinatesClicked() {
+        `when`(media.coordinates).thenReturn(null)
+        `when`(applicationKvStore.getString(LAST_LOCATION)).thenReturn("37.773972,-122.431297")
+        fragment.onUpdateCoordinatesClicked()
```
Since `media` is a mock you can verify if `media.getCoordinates()` is invoked 3 times or not.

You can Assert for Shadow Activity Intent and check if LocationPickerActivity is launched or not.

devarsh-mavani-19 replied:
how can i check if media.getCoordinates is invoked 3 times?

madhurgupta10 replied:
You can simply use `verify(media, times(3)).getCoordinates() `

madhurgupta10 commented on app/src/test/kotlin/fr/free/nrw/commons/media/MediaDetailFragmentUnitTests.kt:
```diff
@@ -229,6 +231,22 @@ class MediaDetailFragmentUnitTests {
+    fun testOnUpdateCoordinatesClickedNotNullValue() {
+        `when`(media.coordinates).thenReturn(LatLng(-0.000001, -0.999999, 0f))
+        `when`(applicationKvStore.getString(LAST_LOCATION)).thenReturn("37.773972,-122.431297")
+        fragment.onUpdateCoordinatesClicked()
```
Here you can verify if `media.getCoordinates()` is invoked only 1 time or not.

You can also Assert for Shadow Activity Intent and check if LocationPickerActivity is launched or not.

madhurgupta10 commented:
@devarsh-mavani-19 Can you also sync your branch with the latest master?

devarsh-mavani-19 commented:
> @devarsh-mavani-19 Can you also sync your branch with the latest master?

yes sure. Do i have to run cherry pick command? I am not sure how can i sync this branch with master. Can you please help me? @madhurgupta10

madhurgupta10 commented:
> > @devarsh-mavani-19 Can you also sync your branch with the latest master?
> 
> yes sure. Do i have to run cherry pick command? I am not sure how can i sync this branch with master. Can you please help me? @madhurgupta10

First set up an upstream to `https://github.com/commons-app/apps-android-commons`  then either `pull upstream master` or `fetch` and `merge`

devarsh-mavani-19 commented:
> > > @devarsh-mavani-19 Can you also sync your branch with the latest master?
> > 
> > 
> > yes sure. Do i have to run cherry pick command? I am not sure how can i sync this branch with master. Can you please help me? @madhurgupta10
> 
> First set up an upstream to `https://github.com/commons-app/apps-android-commons` then either `pull upstream master` or `fetch` and `merge`
> 
> Link for help - https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork

Thanks for resource

devarsh-mavani-19 commented:
@nicolas-raoul I have added all the tests and it is passing all the tests. I have also implemented changes suggested by @madhurgupta10. Can you please review the latest commit? Thanks

devarsh-mavani-19 commented:
Are any changes required? Or can I work on different issues? @nicolas-raoul

nicolas-raoul commented:
I will merge as soon as Madhur approves (or Madhur can merge).
Yes, feel free to start working on another issue if you want, thanks! :-)
